Add fallback user story fetching logic

tuesmiddt opened the pull request:
Fixes #179 

When logic in #173 causes no story to be found, `story["allAttemped"]` will be `true` and the following logic will be used to find a story:
  - Filter by:
    - assessment is published
    - assessment story is not nil
    - assessment `open_at` is before now and `close_at` is after now
    - **NO** submission status filtering
  - Sort by:
    - Assessment `close_at`, descending
    - Assessment type (path>mission>sidequest>contest) for assessments that close at the same time

@ningyuansg Note that the `story` field may be `{story: null, allAttempted: true}` in the event that both methods fail to find a qualifying story. At the moment, the only time I see this happening is at the start of semester before any assessment is open.

indocomsoft commented on lib/cadet/assessments/assessments.ex:
```diff
@@ -29,20 +29,45 @@ defmodule Cadet.Assessments do
+      case direction do
+        :forward ->
+          query
+          |> where([_, s], is_nil(s.id) or s.status == "attempting")
```
Prefer atom `:attempting`

indocomsoft replied:
Also, if I read #179 correctly, `all_attempted === false` if and only if there is any open assessment whose submission's status is `unattempted` (which in our case is basically the `is_nil(s.id)`). I don't think `:attempting` should be here?

tuesmiddt replied:
1. Atoms don't work here. I think the `where` macro just expands it into some raw SQL query (so ecto_enum library is not involved). All our enums as of now should be queried by their string value.

2. In the call, he said that the `all_attempted` flag is essentially to tell whether the first or second kind of logic is used. I personally feel like `attempting` should be treated the same as `not_attempted` in the sense that the assessment has not been "completed".

@ningyuansg Can you confirm whether `allAttempted` flag should be true or false when status is `attempting`?

tuesmiddt replied:
Clarified with him. It should be `false`.

indocomsoft replied:
Alright, cool.

indocomsoft commented on lib/cadet/assessments/assessments.ex:
```diff
@@ -29,20 +29,45 @@ defmodule Cadet.Assessments do
+      |> Multi.run(:default, fn _ -> {:ok, get_user_story_by_direction(user)} end)
+      |> Multi.run(:result, fn %{default: default_res} ->
+        if default_res,
+          do: {:ok, %{all_attempted: false, story: default_res}},
```
If it ends up multi-line, you should use the syntactic sugar
```elixir
if default_res do
  {:ok, %{all_attempted: false, story: default_res}}
else
  {:ok, %{all_attempted: true, story: get_user_story_by_direction(user, :backward)}}
end

tuesmiddt replied:
Fixed

indocomsoft commented on lib/cadet/assessments/assessments.ex:
```diff
@@ -29,20 +29,45 @@ defmodule Cadet.Assessments do
+    story
+  end
+
+  @spec get_user_story_by_direction(%User{}, :forward | :backward) :: String.t() | nil
```
Tbh I feel like `:forward` isn't a good enough description since you add another filtering as compared to `:backward`

tuesmiddt replied:
Renamed it. Is this better?

indocomsoft commented on test/cadet_web/controllers/user_controller_test.exs:
```diff
@@ -60,13 +65,18 @@ defmodule CadetWeb.UserControllerTest do
 
-      assert story == Enum.fetch!(assessments, 1).story
+      expected_story = %{
+        "story" => Enum.fetch!(assessments, 1).story,
```
Ehh, `assessments` is a tuple right? Then I think you should pattern match on it instead.

tuesmiddt replied:
It's a list and i'm getting the second item so I feel it actually looks worse:

```
      [_, assessment | _] = assessments

      expected_story = %{
        "story" => assessment.story,
        "allAttempted" => false
      }
```

indocomsoft replied:
Oops I derped

indocomsoft replied:
Oh shoot sorry I derped

tuesmiddt commented:
@indocomsoft Fixed business logic and updated tests. Please review again.

@ningyuansg I renamed `allAttempted` to `playStory` with reversed values from before i.e. story belongs to `not_attempted` assessment -> `playStory: true` and vice versa for clarity
